storage/remote/queue_manager.go

```diff
@@ -737,26 +812,70 @@ func (s *shards) stop() {
 	}
 }
 
-// enqueue a sample.  If we are currently in the process of shutting down or resharding,
+// bufferPendingSamples listens for incoming pending samples that are caused
+// due to closing of queues. They are stored in a temporary buffer before
+// being appended to the new shards.
+func (s *shards) bufferPendingSamples() {
+	var (
+		wg             sync.WaitGroup
+		bufLock        sync.Mutex
+		buffer         []sample
+		oldShardsCount = len(*s.queueBuf)
+	)
+	if oldShardsCount == 0 {
+		return
+	}
+	wg.Add(oldShardsCount)
+	for i := 0; i < oldShardsCount; i++ {
+		go func(i int) {
+			buf := *s.queueBuf
+			temp, ok := <-buf[i]
+			if !ok {
+				wg.Done()
+				return
+			}
+			bufLock.Lock()
+			buffer = append(buffer, temp...)
+			bufLock.Unlock()
+			wg.Done()
+		}(i)
+	}
+	wg.Wait()
+	sort.SliceStable(buffer, func(i, j int) bool {
+		return buffer[i].t < buffer[j].t
+	})
+	s.qm.RedirectPrevSamples(buffer)
+}
+
+// updateQueues replaces the old queues with the new ones. Post replacement,
+// the blocking of incoming samples is released by storing false in newQueuesAvailable.
+func (s *shards) updateBuffers() {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	s.queueBuf = s.newQueueBuf
+	s.newQueueBuf = nil
+	s.newQueuesAvailable.Store(false)
+}
+
+// enqueue a sample. If we are currently in the process of shutting down or resharding,
 // will return false; in this case, you should back off and retry.
-func (s *shards) enqueue(ref uint64, sample sample) bool {
+func (s *shards) enqueue(ref uint64, sample sample, isPending bool) bool {
 	s.mtx.RLock()
 	defer s.mtx.RUnlock()
 
-	select {
-	case <-s.softShutdown:
+	if isPending {
+		shard := uint64(ref) % uint64(len(s.queues))
+		s.queues[shard] <- sample
+		return true
+	}
+	if s.newQueuesAvailable.Load() {
```

Comment:
The name of this variable is a bit hard to follow when it's being used like this. As far as I can tell you're only using it here, and only to block enqueuing of new samples if we're resharding.

Reply:
Sure, I think, renaming to `blockIncomingSamples` along with a descriptive comment should be good.